illumos-utils/src/running_zone.rs

```diff
@@ -154,26 +250,66 @@ impl RunningZone {
     }
 
     /// Runs a command within the Zone, return the output.
+    #[cfg(target_os = "illumos")]
     pub fn run_cmd<I, S>(&self, args: I) -> Result<String, RunCommandError>
     where
         I: IntoIterator<Item = S>,
         S: AsRef<std::ffi::OsStr>,
     {
+        use std::os::unix::process::CommandExt;
+        let id = self.id.expect("Must have a zone ID for a running zone");
+        let template =
+            std::sync::Arc::new(zenter::Template::new().map_err(|err| {
```

Comment:
I don't really understand the template things going on, but want to double check: all of the raw calls in `zenter::Template::new()` affect our _current_ process, right? Do we have to worry about thread safety or panic safety here? The thread safety concern specifically being - can what we're doing in `Template::new()` mess up some other thread that might be doing arbitrary things, including potentially calling `run_cmd` on this or some other zone? And panic safety: if we panic between this line and `template.clear()`, `template`'s drop will close the fd but won't clear the template - is that okay?

Reply:
> all of the raw calls in zenter::Template::new() affect our current process, right?

My understanding is that they affect the current _thread_. From `man ct_tmpl_clear`:

> The `ct_tmpl_clear()` function clears calling thread's active template.

So I think if another thread is calling `run_cmd` on another zone, that should not interfere. The calls will set that thread's template, independently from this one. If it's the same zone, I think the same thing applies. All this is synchronous, so this shouldn't be swapped around onto different threads, as might happen if it were a `tokio` task.

But now I'm worried about `tokio`...That should be OK, though, right? What I mean is, there is no await point inside `run_cmd`, which means that the task will run that method to completion without yielding. It'll not be swapped out part-way through, where another call to `run_cmd()` might activate / clear a different template on the same OS thread. Does that sound right? It makes me nervous nonetheless.

> And panic safety: if we panic between this line and template.clear(), template's drop will close the fd but won't clear the template - is that okay?

I think so, again because the template is activated for a thread, not the process. (We also abort rather than unwind, but I'm not sure that's relevant to your question.)

Reply:
> But now I'm worried about tokio...That should be OK, though, right? What I mean is, there is no await point inside run_cmd, which means that the task will run that method to completion without yielding. It'll not be swapped out part-way through, where another call to run_cmd() might activate / clear a different template on the same OS thread. Does that sound right? It makes me nervous nonetheless.

I think this is right; in particular, `run_cmd` is not even an async function, so there's no opportunity for tokio to move it around. It might be worth adding a comment noting that it's important that this function continues to be non-async because it does illumos-specific thread-local things that would not play nicely with tokio.

> I think so, again because the template is activated for a thread, not the process. (We also abort rather than unwind, but I'm not sure that's relevant to your question.)

Abort instead of unwind is definitely relevant, because if we abort we don't have to worry about panic safety at all. If we want this to be panic-safe from an unwind point of view, I'm not sure it is: if I'm on a thread that calls this function, but inside (possibly deeply nested inside) a [catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html), it's possible a panic here could leave my thread around with this template activated. That's a lot of "if"s that don't apply in practice, so maybe that is also worth just commenting? The real fix would be to have the `clear` called in a `Drop` method, guaranteeing it would happen even if we unwound.

Reply:
Done in bbdffe8f1